### ISSUE TITLE
"Correct" parsing of f-strings

### ISSUE THREAD

**Issue by CoffeeTableEspresso** [python/cpython#104117]
# Feature or enhancement

f-strings currently may not contain the same quotes inside any expressions.

e.g. `f" dsadsads { "dsadasdsa" } dsadsadsa"` is invalid, since we used double-quotes for both the f-string and in the expression in the f-string.

If you want to use a string literal in an f-string currently, you must use the other kind of quotes, in order for Python to parse it correctly. e.g. You must do `f" dsadsads { 'dsadasdsa' } dsadsadsa"` rather than the previous version.

# Pitch

This restriction should not be necessary. There is only one correct way to interpret the above f-strings. The only reason that it is not allowed is because of how the CPython parser handles f-strings, not because this (or any other example) is ambiguous.

# Prior Art

Most other languages with interpolated strings do not have the same restriction as Python, which suggests that this restriction should be removed.

**Comment by JelleZijlstra** [comment]
You're in luck, this feature is already implemented in the upcoming Python 3.12. See PEP-701 for details.

```
% ./python.exe 
Python 3.12.0a7+ (heads/tvobject-dirty:f4ea500baf, May  2 2023, 19:07:46) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> f" dsadsads { "dsadasdsa" } dsadsadsa" 
' dsadsads dsadasdsa dsadsadsa'
>>> 
```

**Comment by CoffeeTableEspresso** [comment]
Thanks for the quick reply, didn't realize someone had already worked on this! Glad to see it.
